### ISSUE TITLE
[HELP WANTED/TESTING] Testing and finalizing game image installer functionality

### ISSUE THREAD

**Issue by d0k3** [d0k3/GodMode9#640]
This is continued from #618 and #621. All the tests mentioned refer these two pull requests. Thanks @TurdPooCharger for all your help here, now we may need some additional help. This is the current state of things:

**1. Is that stuff fixed for good?**
I'm just taking over the list here. I *think* everything here is fixed, but a second look can't hurt.

* bad CMD file without CMACs (see test 1): Should be fixed for good. Only system CMDs have no CMACs.
* `title.db` entry, bytes 0x0C...0x0 (see test 1): "title version" - takes over NCCH title version too for NCCH with DLP only (thanks @ihaveamac)
* cleanup of orphaned files from earlier installs (see test 2): Before an install, we wipe the `content/data` folder now. Existing saves with the correct size are left untouched.
* ticket device ID is always set to zero (see test 3): We keep the device ID if it matches the system now. It is zeroed only where required.
* savegames for TWL are not auto-created (see test 4): This is implemented now. The save files won't match system created ones exactly (in fact, only the first 0x200 byte should be identical for `public.sav` and `private.sav`), but it should still work. We are unsure if `banner.sav` also work. Further testing is required here.
* `title.db` entry, bytes 0x20...0x23 (see test 5): "extdata id low" - this is now properly taken over
* installing games and handling seeds (see test 7): GM9 now does import seeds from an existing `seeddb.bin` when installing a seed-encrypted CIA or TMD game image. If no matching seed is found, we add a titletag entry so the 3DS OS can download the seed on itself.

**2. Stuff that may need further looking into (not high priority)**
Nothing here has high priority. Just a quick reminder.

* `title.db` entry, bytes 0x00...0x03 and 0x30...0x3F for `DS INTERNET` (see test 3) - we have no clue what happens here. Values seem random.
* Difference in ticket between CIA install and .3DS install (see test 5): No clue what happens here yet, but no effect on functionality. Maybe @aspargas2 has a better understanding of how this happens than me.
* installing N3DS exclusives on O3DS (see test 8): I personally don't think that's an issue. GM9 does what the user wants after all. I'd be okay with putting out a warning (if we're able to recognise N3DS exclusives properly), but I don't think it's required.

**3. Stuff that must be looked into**
I need to have a closer look here, and there's also some open questions (which I'll add to the list).

* update CIAs creating save files (see test 8): Some update CIAs have the save flag set, thus GM9 creates a save file for them (which it should not). I need ideas on how to solve this. Should we disable generating save games for update CIAs in general? And can we properly recognise update CIAs?
* leftovers when installing system titles (see test 10): Oooof. Tbh I'm not really sure I understand what happens there. You were able to restore the system titles from GM9, then, after the update, you noticed there were some leftover titles (older versions), right? There may actually be no solution to this, as the problem seems to be in the 3DS OS update and we do not completely know how this works. May be related to TMD numbering (which defaults to 1 in GM9). Does anyone think this needs work?

**4. Archive / Done (meaning: fixed for good**
This list includes everything that has been fixed and checked. It's just a copy of the former text from above.

* handling corrupt game images (see test 12): GM9 does no more allow to install corrupt CIAs or TMDs. You can't build a CIA from a corrupt TMD anymore either. Don't like it? You can discuss here. I personally match @TurdPooCharger's opinion and think this is the right way to do it.

**Comment by TurdPooCharger** [comment]
To get back into bug testing mode, I'll like to comment on this one topic to get the ball rolling again.

> leftovers when installing system titles (see test 10): Oooof. Tbh I'm not really sure I understand what happens there. You were > able to restore the system titles from GM9, then, after the update, you noticed there were some leftover titles (older versions), > right? There may actually be no solution to this, as the problem seems to be in the 3DS OS update and we do not completely > > know how this works. May be related to TMD numbering (which defaults to 1 in GM9). Does anyone think this needs work?
This orphaning of system titles' files must have been present since the original Safehax/2.1.0 downgrade and Decrypt9 CTRTransfers. I don't believe you and the other early pioneer devs could have foreseen this quirk when repairing firmware softbricks was still being studied.

The really picky person might not like having older leftover system titles eating up of some their ctrnand space. There's no harm or impedance to the functionalities of the repaired 3DS firmware. I think we shouldn't put too much stock in this nonissue in regards to GameInstaller.

If anyone would like to revisit this later, I would like to share a rough idea of using a custom gm9 script as a post clean-up for CTRTransfer + system update. Sort and removes orphaned files by looking for and separating files listed in the title entries `title.db`, `*.cmd`, and `*.tmd`.

**Comment by d0k3** [comment]
Glad to have you back, @TurdPooCharger !

Regarding that issue, I agree. It's not that important for now. Personally I think the leftovers have something to do with the CMD/TMD filenames (the systems uses something specific there, but we don't know what it is based on).

**Comment by TurdPooCharger** [comment]
> handling corrupt game images (see test 12): GM9 does no more allow to install corrupt CIAs or TMDs. You can't build a CIA from a corrupt TMD anymore either. Don't like it? You can discuss here. I personally match @TurdPooCharger's opinion and think this is the right way to do it.
- Test build `gameinstaller_finalize` compiled with last commit `23b4b50`.

## Test No. 13 - A few revisits

### 13A. Does GameInstaller (now) rejects installing bad CIAs?
Yes.

I attempted to install the purposely edit `FBI (bad).cia` from **Test No. 12**. **GameInstaller** reported back this error.
- > Install failed

To make sure this wasn't a fluke, I modified a GM9 converted .3DS → .CIA (CRC32: **1A599210**, decrypted) of **Pokémon Sun** from **Test No. 5** by hex editing the very byte of the game's manual .APP content from **00** → **FF**.

I could not trick **GameInstaller**. It _almost_ installs the game BUT again returns the _Install failed_ error message.

:+1: 

### 13B. Does GodMode9 (now) rejects building bad CIAs?
Yes.

I install clean copies of **FBI** and **Pokémon Sun** and hex edit the flaws to those .APP as describes in **13A**. **GodMode9** both times returns _CIA build failed_.

:+1: 

### 13C. Are seeds importable?
No.

- **sdmc:/gm9/support/**`seeddb.bin` is put in place.
- **1:/data/**<**ID0**>**/sysdata/0001000f/**`00000000` is blank/default.
- From **Test No. 7**, _Bloodstained: Curse of the Moon (JPN)_ is GM9 installed.
- The game's e-manual still crashes.

I checked the **0001000f/**`00000000` → `partitionA.bin` post install. The below modifications were not observed.
- 0x08 bytes TitleID string at offset 0x4000. 
- 0x10 bytes seed string at offset 0x7E80.

### 13D. Are titles uninstallable?
Yes.

1. **A:/dbs/**`title.db` → **Mount as DIFF image** → ****<**A**>** yes**.
2. For the **Pokémon Sun** from **13A**, **T:/**`0004000000164800`
3. **Manage Title...** → **Uninstall title** → **Keep ticket & savegame**

:+1: 

### Bugs

**Keep ticket & savegame** deletes the save. 
- **A:/title/00040000/<TID_LOW>/data/`00000001.sav`

When deciding to keep the .SAV if present, there should be a special condition where the .SAV is deleted if it is uninitialized (first 0x20 bytes are all zeros). 

The uninstalled title's ticket is kept in place, so no problem there.

The **<TID_LOW>** folder is orphaned. It should be removed when no .SAV is kept.

***

When there is a valid .SAV in place (ie, the first 0x20 bytes are not zeros),

- A:/title/00040000/<TID_LOW>/data/**`00000001.sav`

, where the file is by itself or has its game already installed, reinstalling that same game  CIA will overwrite or erase that save.

**Comment by d0k3** [comment]
Thanks a lot, @TurdPooCharger !

For the title manager thingy, I guess you'll like what I've been up to recently, here's a test build:
https://f.secretalgorithm.com/c500P/godmode9-v1.9.2pre1-20-g57fb0675-20201125230935.zip
See: drive menu, HOME menu

I also included a fix for savegames being deleted with the "Keep ticket & savegame option". It's a good idea to actually check those saves and to delete uninitialised ones, however, the implemention is a bit complicated in the current state. I think we can leave that decision in the users hands, anyways. I also edited the op with your new results.

Importing seeds should work. Is there any chance you had the seed save mounted when you tried to install? I'm also somewhat surprise the manual doesn't work (does the main game?), so this may be an issue specific to this game.
